### PR TITLE
Added reader for new config mode

### PR DESCRIPTION
**Pull request by duncanp-sonar**
With this change, the _IConfigurationProvider_ will correctly detect whether it is in standalone, legacy or new connected mode.

There isn't any way to configure a project to be in "new" connected mode yet except by manually setting it up i.e. writing the expected file to disk in the correct location.

Related to #508

### CODE REVIEW

File: src/Integration.UnitTests/NewConnectedMode/BindingConfigurationTests.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+using System;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarLint.VisualStudio.Integration.NewConnectedMode;
+using SonarLint.VisualStudio.Integration.Persistence;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests
+{
+    [TestClass]
+    public class BindingConfigurationTests
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/NewConnectedMode/BindingConfigurationTests.cs]: Added tests for _BindingConfiguration_ class added yesterday.

File: src/Integration.UnitTests/NewConnectedMode/ConfigurationProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarLint.VisualStudio.Integration.NewConnectedMode;
+using SonarLint.VisualStudio.Integration.Persistence;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/NewConnectedMode/ConfigurationProviderTests.cs]: Tests that the _ConfigurationProvider_ handles legacy and new modes, and prefers legacy to new if both exist (which should never happen in practice).

File: src/Integration.UnitTests/Persistence/SolutionBindingSerializerTests.cs [hunk 3]
```diff
@@ -356,7 +354,7 @@ public void SolutionBindingSerializer_WriteSolutionBinding_AddConfigFileToSoluti
 
         private SolutionBindingSerializer CreateTestSubject()
         {
-            return new SolutionBindingSerializer(this.serviceProvider, this.store, new SonarLintOutputLogger(serviceProvider));
+            return new SolutionBindingSerializer(this.serviceProvider, this.store, new SonarLintOutputLogger(serviceProvider), new FileWrapper());
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Persistence/SolutionBindingSerializerTests.cs]: I extracted the existing code to read from a config file to a helper class. To make it more testable, I changed the extract code to use the _IFile_ abstraction, which meant I had to make this class use the _FileWrapper_ too.
However, I haven't changed the existing _ISolutionBindingSerializer_ tests that actually read and write files from disk - low ROI.

File: src/Integration/MefServices/VsSessionHost.cs [hunk 4]
```diff
@@ -295,7 +296,12 @@ private void RegisterLocalServices()
             this.localServices.Add(typeof(ISolutionRuleSetsInformationProvider), new Lazy<ILocalService>(() => new SolutionRuleSetsInformationProvider(this, Logger)));
             this.localServices.Add(typeof(IRuleSetSerializer), new Lazy<ILocalService>(() => new RuleSetSerializer(this)));
             this.localServices.Add(typeof(ISolutionBindingSerializer), new Lazy<ILocalService>(() => new SolutionBindingSerializer(this)));
-            this.localServices.Add(typeof(IConfigurationProvider), new Lazy<ILocalService>(() => new LegacyConfigurationProviderAdapter(this.GetService<ISolutionBindingSerializer>())));
+            this.localServices.Add(typeof(IConfigurationProvider), new Lazy<ILocalService>(() =>
+            {
+                var solution = this.GetService<SVsSolution, IVsSolution>();
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/VsSessionHost.cs]: I decided to make the lazy service constructor do the work of fetching the _IVsSolution_. The alternative would have been to pass an _IServiceProvider_ to the _ConfigurationProvider_.

File: src/Integration/NewConnectedMode/ConfigFileUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+using SonarLint.VisualStudio.Integration.Helpers;
+using SonarLint.VisualStudio.Integration.Persistence;
+using SonarLint.VisualStudio.Integration.Resources;
+using SonarQube.Client.Helpers;
+
+namespace SonarLint.VisualStudio.Integration.NewConnectedMode
+{
+    internal static class ConfigFileUtilities
+    {
+        public static BoundSonarQubeProject ReadBindingFile(string fullConfigFilePath, ICredentialStore credentialStore, ILogger logger, IFile fileWrapper)
```
Comment (duncanp-sonar) [inline src/Integration/NewConnectedMode/ConfigFileUtilities.cs]: Code extracted from _SolutionBindingSerializer_. Minor changes to make it use the _IFile_ abstraction.

File: src/Integration/NewConnectedMode/ConfigurationSerializer.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            }
+            if (credentialStore == null)
+            {
+                throw new ArgumentNullException(nameof(credentialStore));
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            Debug.Assert(fileWrapper != null);
```
Comment (duncanp-sonar) [inline src/Integration/NewConnectedMode/ConfigurationSerializer.cs]: I'm doing null checks to the arguments to public constructor, but just asserting for the other argument.

File: src/Integration/NewConnectedMode/ConfigurationSerializer.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+using SonarLint.VisualStudio.Integration.Helpers;
+using SonarLint.VisualStudio.Integration.Persistence;
+
+namespace SonarLint.VisualStudio.Integration.NewConnectedMode
+{
+    /// <summary>
+    /// Reads/write the binding configuration for the new connected mode.
+    /// The legacy connected mode binding is written using <see cref="SolutionBindingSerializer"/>.
+    /// </summary>
+    internal class ConfigurationSerializer : ISolutionBindingSerializer
```
Comment (duncanp-sonar) [inline src/Integration/NewConnectedMode/ConfigurationSerializer.cs]: I decided to reuse the existing _ISolutionBindingSerializer_ to read/write the new config.
Currently only the _read_ part is implemented, and it looks for a file called _.sonarlint\\[SOLUTIONFILENAME].sqconfig_

File: src/Integration/NewConnectedMode/ConfigFileUtilities.cs (hunk 5 above)
Comment (Evangelink) [reply src/Integration/NewConnectedMode/ConfigFileUtilities.cs]: [cosmetic] I would split this long line because it is hard to read/review.

**Pull request merged**